pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+		config.newLabels = labels
+	}
+}
+
+// Configure sets the owner reference and merges the labels. Other options modify the logic
+// of apply function and therefore need to be checked manually.
+func (c *ssaApplyObjectConfiguration) Configure(obj client.Object, s *runtime.Scheme) error {
+	if c.owner != nil {
+		if err := controllerutil.SetControllerReference(c.owner, obj, s); err != nil {
+			return err
+		}
+	}
+	MergeLabels(obj, c.newLabels)
+
+	return nil
+}
+
+// ApplyObject creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then when the spec content has changed (based on the content of the annotation in the original object) then it
+// is automatically updated. If it looks to be same then based on the value of forceUpdate param it updates the object or not.
+// The return boolean says if the object was either created or updated (`true`). If nothing changed (ie, the generation was not
+// incremented by the server), then it returns `false`.
+//
+// NOTE: the return boolean IS ALWAYS TRUE on success by default. This is much more efficient. If you truly need
+// to determine whether the object has been updated or not (which is not the case anywhere in the codebase but the tests),
+// you can use the DetermineUpdate() option.
+func (c *SsaApplyClient) ApplyObject(ctx context.Context, obj client.Object, options ...SsaApplyObjectOption) (bool, error) {
+	config := newSsaApplyObjectConfiguration(options...)
+	if err := config.Configure(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if err := prepareForSSA(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if config.skipIf != nil && config.skipIf(obj) {
+		return false, nil
+	}
+
+	updated := true
+
+	// NOTE: once the SSA migration is not needed anymore, read the orig conditionally only when config.determineUpdate is true.
+
+	// this cannot fail - we would have failed previously already because the scheme wouldn't have known the object's kind
+	origRo, _ := c.Client.Scheme().New(obj.GetObjectKind().GroupVersionKind())
+	orig := origRo.(client.Object)
+
+	if err := c.Client.Get(ctx, client.ObjectKeyFromObject(obj), orig); err != nil && !apierrors.IsNotFound(err) {
+		return false, err
+	}
+
+	// NOTE: remove this once this code is used for all "apply-style" functionality and has updated all the objects in the cluster.
+	if isSsaMigrationNeeded(orig, c.NonSSAFieldOwner) {
+		if err := c.MigrateToSSA(ctx, orig); err != nil {
+			return false, err
+		}
+	}
+
+	if err := c.Client.Patch(ctx, obj, client.Apply, client.FieldOwner(c.FieldOwner), client.ForceOwnership); err != nil {
+		return false, fmt.Errorf("unable to patch '%s' called '%s' in namespace '%s': %w", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), obj.GetNamespace(), err)
+	}
+
+	if config.determineUpdate {
+		updated = obj.GetGeneration() != orig.GetGeneration()
+	}
```

Comment:
we could also check labels & annotations

Reply:
I'm hoping we will drop this check altogether and will just not bother trying to determine if there was an update or not. But I kept this here to lower the impact on the tests that do work with this value. In all other places that I am aware, the result of this is just used for logging purposes.

Reply:
Yeah, that's what I wanted to propose actually, if there is no real usage of this returned value, then I would simply drop it.

Reply:
Makes sense, let's just bite the bullet and keep the production code as simple as possible.

Reply:
:+1: